microprofile-fault-tolerance/src/test/java/org/jboss/eap/qe/microprofile/fault/tolerance/v20/FaultTolerance20AsyncTest.java

```
@@ -0,0 +1,295 @@
+        expectedResponses.put("Hello from @Bulkhead @Timeout @Retry method", 0);
+        expectedResponses.put("Fallback Hello", 40);
+
+        testBulkhead(40, baseApplicationUrl + "async?operation=bulkhead-timeout-retry&fail=true", expectedResponses);
+    }
+
+    private static void testBulkhead(int parallelRequests, String url, Map<String, Integer> expectedResponses)
+            throws InterruptedException {
+        Set<String> violations = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        Queue<String> seenResponses = new ConcurrentLinkedQueue<>();
+
+        ExecutorService executor = Executors.newFixedThreadPool(parallelRequests);
+        for (int i = 0; i < parallelRequests; i++) {
+            executor.submit(() -> {
+                try {
+                    seenResponses.add(get(url).asString());
+                } catch (Exception e) {
+                    violations.add("Unexpected exception: " + e.getMessage());
+                }
+            });
+        }
+        executor.shutdown();
+        boolean finished = executor.awaitTermination(10, TimeUnit.SECONDS);
+        assertThat(finished, is(true));
+
+        for (String seenResponse : seenResponses) {
+            if (!expectedResponses.containsKey(seenResponse)) {
+                violations.add("Unexpected response: " + seenResponse);
+            }
+        }
+        for (Map.Entry<String, Integer> expectedResponse : expectedResponses.entrySet()) {
+            int count = 0;
+            for (String seenResponse : seenResponses) {
+                if (expectedResponse.getKey().equals(seenResponse)) {
+                    count++;
+                }
+            }
+            if (count != expectedResponse.getValue()) {
+                violations.add("Expected to see " + expectedResponse.getValue() + " occurrence(s) but seen " + count
+                        + ": " + expectedResponse.getKey());
+            }
+        }
+        assertThat(violations, is(empty()));
```

Same as previously - why not fail fast?

---

Same as above.